fontforgeexe/shapers/harfbuzz.cpp

```diff
@@ -0,0 +1,349 @@
+/* Copyright 2024 Maxim Iorsh <iorsh@users.sourceforge.net>
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ * Redistributions of source code must retain the above copyright notice, this
+ * list of conditions and the following disclaimer.
+ * Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ * The name of the author may not be used to endorse or promote products
+ * derived from this software without specific prior written permission.
+ * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR IMPLIED
+ * WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO
+ * EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
+ * OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+ * WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
+ * OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
+ * ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+#include "harfbuzz.hpp"
+
+#include <algorithm>
+#include <cassert>
+
+extern "C" {
+#include "gfile.h"
+#include "splinechar.h"
+#include "splinefont_enums.h"
+#include "utype.h"
+}
+
+namespace ff::shapers {
+
+HarfBuzzShaper::HarfBuzzShaper(std::shared_ptr<ShaperContext> context)
+    : context_(context) {
+    FILE* ttf_file = GFileTmpfile();
+
+    _WriteTTFFont(
+        ttf_file, context_->sf, ff_ttf, NULL, bf_ttf,
+        ttf_flag_otmode | ttf_flag_oldkernmappedonly | ttf_flag_fake_map,
+        context_->get_enc_map(context_->sf), ly_fore);
```

Comment:
You can probably speed things up by skipping generating tables that HarfBuzz does not use, e.g. glyph outlines (glyf and CFF table) which is usually the larges font tables and the slowest to build.